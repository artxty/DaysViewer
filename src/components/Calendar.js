import React from 'react';
import PropTypes from 'prop-types';
import { InteractionManager } from 'react-native';
import { CalendarList } from 'react-native-calendars';
import iosColors from 'ios-colors';

import {
  isAfter,
  getDateId,
  getMonthId,
  getDateFromDateId,
} from '../helpers/date';
import { RATING_COLOR_MAP } from '../constants/Colors';

export default function Calendar({
  days = {}, onScroll, onDayPress, pastMonthsRange,
}) {
  /* TODO: is there any other solution? */
  function getMarkedDates() {
    return Object.keys(days).reduce((formattedDays, dayId) => {
      const dayDate = getDateFromDateId(dayId);
      const currentDate = new Date();
      if (isAfter(dayDate, currentDate)) {
        return {
          ...formattedDays,
          [dayId]: {
            disableTouchEvent: true,
          },
        };
      }

      const rating = days[dayId] !== null ? days[dayId].rating : 0;
      const note = days[dayId] && days[dayId].note;
      return {
        ...formattedDays,
        [dayId]: {
          marked: !!note,
          dotColor: iosColors.systemGray.light,
          selected: true,
          selectedColor: RATING_COLOR_MAP[rating],
        },
      };
    }, {});
  }

  function handleVisibleMonthsChange(months = []) {
    // TODO: add two-three extra months to hide loading?
    const visibleMonthsIds = months.map(({ year, month }) => {
      const monthDate = new Date(year, month - 1);
      return getMonthId(monthDate);
    });
    /*
      InteractionManager.runAfterInteractions is necessary here
      to solve an issue of months displaying:
      https://github.com/wix/react-native-calendars/issues/884
    */
    return InteractionManager.runAfterInteractions(() => onScroll(visibleMonthsIds));
  }

  function handleDayPress({ year, month, day }) {
    const dayDate = new Date(year, month - 1, day);
    const dayId = getDateId(dayDate);
    onDayPress(dayId);
  }

  return (
    <CalendarList
      /*
        Changing pastScrollRange dynamically doesn't re-render the component
        to force re-render 'key' prop need to be updated
        */
      key={pastMonthsRange}
      markedDates={getMarkedDates()}
      onVisibleMonthsChange={handleVisibleMonthsChange}
      onDayPress={handleDayPress}
      pastScrollRange={pastMonthsRange}
      futureScrollRange={0}
      displayLoadingIndicator
    />
  );
}

Calendar.propTypes = {
  onScroll: PropTypes.func.isRequired,
  onDayPress: PropTypes.func.isRequired,
  days: PropTypes.shape({
    rating: PropTypes.number,
    note: PropTypes.string,
  }).isRequired,
  pastMonthsRange: PropTypes.number.isRequired,
};
