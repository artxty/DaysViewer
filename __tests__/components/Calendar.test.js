import React from 'react';
import renderer from 'react-test-renderer';

import Calendar from '../../src/components/Calendar';

jest.mock('react-native-calendars', () => ({
  CalendarList: 'CalendarList',
}));

const pastMonthsRange = 3;

describe('Calendar', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <Calendar
        pastMonthsRange={pastMonthsRange}
        days={{}}
        onDayPress={jest.fn()}
        onScroll={jest.fn()}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('transforms days prop into markedDates correctly', () => {
    const possibleRatings = [0, 1, 2, 3, 4, 5];
    const rawDays = possibleRatings.reduce((acc, r) => ({
      ...acc,
      [`mock day id ${r}`]: {
        rating: r,
        note: 'some note',
      },
    }), {});

    const tree = renderer.create(
      <Calendar
        pastMonthsRange={pastMonthsRange}
        days={rawDays}
        onDayPress={jest.fn()}
        onScroll={jest.fn()}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('calls onScroll when CalendarList calls onVisibleMonthsChange', async () => {
    const onScrollMock = jest.fn();
    const tree = renderer.create(
      <Calendar
        pastMonthsRange={pastMonthsRange}
        days={{}}
        onDayPress={jest.fn()}
        onScroll={onScrollMock}
      />,
    );

    const testInstance = tree.root;
    const calendarList = testInstance.find((el) => el.type === 'CalendarList');
    await calendarList.props.onVisibleMonthsChange([]);

    expect(onScrollMock).toBeCalled();
  });
});
