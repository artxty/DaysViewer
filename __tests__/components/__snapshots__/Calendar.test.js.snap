// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar renders correctly 1`] = `
<CalendarList
  displayLoadingIndicator={true}
  futureScrollRange={0}
  markedDates={Object {}}
  onDayPress={[Function]}
  onVisibleMonthsChange={[Function]}
  pastScrollRange={3}
/>
`;

exports[`Calendar transforms days prop into markedDates correctly 1`] = `
<CalendarList
  displayLoadingIndicator={true}
  futureScrollRange={0}
  markedDates={
    Object {
      "mock day id 0": Object {
        "dotColor": "rgb(142, 142, 147)",
        "marked": true,
        "selected": true,
        "selectedColor": "rgb(229, 229, 234)",
      },
      "mock day id 1": Object {
        "dotColor": "rgb(142, 142, 147)",
        "marked": true,
        "selected": true,
        "selectedColor": "rgb(255, 59, 48)",
      },
      "mock day id 2": Object {
        "dotColor": "rgb(142, 142, 147)",
        "marked": true,
        "selected": true,
        "selectedColor": "rgb(255, 149, 0)",
      },
      "mock day id 3": Object {
        "dotColor": "rgb(142, 142, 147)",
        "marked": true,
        "selected": true,
        "selectedColor": "rgb(255, 204, 0)",
      },
      "mock day id 4": Object {
        "dotColor": "rgb(142, 142, 147)",
        "marked": true,
        "selected": true,
        "selectedColor": "rgb(164, 202, 80)",
      },
      "mock day id 5": Object {
        "dotColor": "rgb(142, 142, 147)",
        "marked": true,
        "selected": true,
        "selectedColor": "rgb(52, 199, 89)",
      },
    }
  }
  onDayPress={[Function]}
  onVisibleMonthsChange={[Function]}
  pastScrollRange={3}
/>
`;
